### PR TITLE
Feature/187 substitutables strings

### PR DESCRIPTION
**Pull request by DamienSAVOYAT**
Add SubstitutableStringModel class for allow substitute a string with other values.

### CODE REVIEW

File: OpenXMLSDK.Engine/ReportEngine/DataContext/SubstitutableStringModel.cs [hunk 1]
```diff
@@ -0,0 +1,86 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
```
Comment (mathieumack) [inline OpenXMLSDK.Engine/ReportEngine/DataContext/SubstitutableStringModel.cs]: Please remove this double empty line to keep only one

File: OpenXMLSDK.Engine/ReportEngine/DataContext/SubstitutableStringModel.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    public class SubstitutableStringModel : BaseModel
+    {
+        /// <summary>
+        /// List of substitution texts
+        /// </summary>
+        public ContextModel DataSource { get; set; }
+
+        /// <summary>
+        /// Used to define the final rendering string You can set precision of other :
+        /// Ex : '{0:G2} kV' for the value 3.230 will generate '3.23 kV' string
```
Comment (mathieumack) [inline OpenXMLSDK.Engine/ReportEngine/DataContext/SubstitutableStringModel.cs]: Could you update this exemple to use the <example> tag ?
doc : https://docs.microsoft.com/fr-fr/dotnet/csharp/programming-guide/xmldoc/example

File: OpenXMLSDK.Engine/ReportEngine/DataContext/SubstitutableStringModel.cs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        /// </summary>
+        public string RenderPattern { get; set; }
+
+        #region Constructor
+
+        /// <summary>
+        /// Default constructor
+        /// </summary>
+        public SubstitutableStringModel() : this(null, null)
```
Comment (mathieumack) [inline OpenXMLSDK.Engine/ReportEngine/DataContext/SubstitutableStringModel.cs]: usefull for ?

File: OpenXMLSDK.Engine/ReportEngine/DataContext/SubstitutableStringModel.cs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+        /// <summary>
+        /// Create the generated string
+        /// </summary>
+        /// <returns></returns>
+        public string Render(IFormatProvider formatProvider)
+        {
+            var renders = new List<string>();
+
+            foreach (var baseModel in DataSource.Data.Values)
```
Comment (mathieumack) [inline OpenXMLSDK.Engine/ReportEngine/DataContext/SubstitutableStringModel.cs]: Pleasecheck that Datasource field is not null.
Check also that DataSource.Data.Values is not null, and also each value can be null

File: OpenXMLSDK.Engine/ReportEngine/DataContext/SubstitutableStringModel.cs [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+            }
+
+            try
+            {
+                return string.Format(RenderPattern, renders.ToArray());
+            }
+            // When there are less parameters than expected in RenderPattern string 
+            catch (FormatException)
+            {
+                // We add empty parameter
```
Comment (mathieumack) [inline OpenXMLSDK.Engine/ReportEngine/DataContext/SubstitutableStringModel.cs]: To be discussed. Not sure about this behaviour.

File: OpenXMLSDK.Engine/ReportEngine/DataContext/SubstitutableStringModel.cs (hunk 3 above)
Comment (DamienSAVOYAT) [reply OpenXMLSDK.Engine/ReportEngine/DataContext/SubstitutableStringModel.cs]: No, it's useless

File: OpenXMLSDK.Engine/ReportEngine/DataContext/SubstitutableStringModel.cs (hunk 5 above)
Comment (DamienSAVOYAT) [reply OpenXMLSDK.Engine/ReportEngine/DataContext/SubstitutableStringModel.cs]: Ok, after discussion change to return original string.

File: OpenXMLSDK.Engine/ReportEngine/DataContext/SubstitutableStringModel.cs [hunk 6]
```diff
@@ -50,10 +46,15 @@ public SubstitutableStringModel(string renderPattern, ContextModel dataSource) :
...
         public string Render(IFormatProvider formatProvider)
         {
-            var renders = new List<string>();
+            if (DataSource is null || DataSource.Data is null || DataSource.Data.Values is null)
+                return string.Empty;
 
+            var renders = new List<string>();
             foreach (var baseModel in DataSource.Data.Values)
             {
+                if (baseModel is null)
```
Comment (mathieumack) [inline OpenXMLSDK.Engine/ReportEngine/DataContext/SubstitutableStringModel.cs]: Please remove this continue statement by a LinqQuery (.Where(e => e != null) in the foreach

Comment (DamienSAVOYAT) [reply OpenXMLSDK.Engine/ReportEngine/DataContext/SubstitutableStringModel.cs]: Ok, change.

**Pull request merged**